libp2p-tcp-transport/src/lib.rs

```diff
@@ -131,6 +132,35 @@ impl Transport for TcpConfig {
             Err((self, addr))
         }
     }
+
+    fn nat_traversal(&self, server: &Multiaddr, observed: &Multiaddr) -> Option<Multiaddr> {
+        let server_protocols: Vec<_> = server.iter().collect();
+        let observed_protocols: Vec<_> = observed.iter().collect();
+
+        if server_protocols.len() != 2 || observed_protocols.len() != 2 {
+            return None;
+        }
+
+        // Check that `server` is a valid TCP/IP address.
+        match (&server_protocols[0], &server_protocols[1]) {
```

Comment:
what if the order of protocols is different? is it possible? what if there are more protocols than 2?

Reply:
Each library is only responsible for its own supported protocols, so we only handle addresses that match `/ipN/.../tcp/...` here. Although it's debatable, I don't think `/tcp/.../ipN/...` makes sense, so we don't allow it.